infrastructure/derive/src/lib.rs

```diff
@@ -33,3 +37,112 @@ pub fn derive_hashable_ordering(tokens: TokenStream) -> TokenStream {
     };
     gen.into()
 }
+
+/// This macro will provide a Hashable implementation to the a given struct using a Digest implementing Hash function
+/// To use this provide #[derive(Hashable)] to the struct and #[Digest = "<Digest>"] with <Digest> being the included
+/// digest the macro should use to impl Hashable individual fields can be skipped by providing them with:
+/// #[Hashable(Ignore)]
+#[proc_macro_derive(Hashable, attributes(Digest, Hashable))]
+pub fn derive_hashable(tokens: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(tokens as DeriveInput);
+    let object_name = &input.ident;
+    let mut digest = None;
+    for attr in &input.attrs {
+        match attr.interpret_meta().unwrap() {
+            syn::Meta::NameValue(val) => {
+                if val.ident.to_string() == "Digest" {
+                    if let syn::Lit::Str(lit) = &val.lit {
+                        digest = Some(lit.value());
+                    }
+                }
+            },
+            _ => (),
+        };
+    }
+    let item = input.data;
+    let fields_text = handle_fields_for_hashable(&item);
+
+    let digest = digest.expect("Could not find Digest attribute"); // this is for the error, if the Digest was not given, this error message will be displayed
+    let varname = Ident::new(&digest, Span::call_site());
+    let gen = quote! {
+        impl  Hashable for #object_name  {
+            fn hash(&self) -> Vec<u8> {
+                let mut hasher = <#varname>::new();
+                #fields_text
+                hasher.result().to_vec()
+            }
+        }
+    };
+    gen.into()
+}
+
+// this function processes the individual fields of the hashable trait macro: derive_hashable
+fn handle_fields_for_hashable(item: &Data) -> proc_macro2::TokenStream {
+    match item {
+        Data::Struct(ref item) => {
+            match item.fields {
+                Fields::Named(ref fields) => {
+                    let recurse = fields.named.iter().map(|f| {
+                        let mut do_we_ignore_field = false;
+                        for attr in &f.attrs {
+                            match attr.interpret_meta().unwrap() {
```

Comment:
Should this unwrap not be handled differently?

Reply:
This should not break, but in the odd case it does, its only going to break on compile as this is when this code runs, this wont break at runtime